add info attribute to ContextParameterValue

mgv99 opened the pull request:
This new attribute adds the possibility to register extra information about a context parameter value.

For instance, with date-time entities, the info attribute can contain a Map with an entry for each part of the date-time (year, month, day...) indicating if it was explicitly said by the user (true value) or not (false value). The used NER system could fill the missing parts of 'April 4th', (year, hour, minutes and seconds) with the current date-time. But checking the info attribute, the bot developer could know which parts were not present initially (maybe to reconfirm them asking again to the user)

mgv99 committed:
add info attribute to ContextParameterValue

This new attribute adds the possibility to register extra information about a context parameter value.

For instance, with date-time entities, the info attribute can contain a Map with an entry for each part of the date-time (year, month, day...) indicating if it was explicitly said by the user (true value) or not (false value). The used NER system could fill the missing parts of 'April 4th', (year, hour, minutes and seconds) with the current date-time. But checking the info attribute, the bot developer could know which parts were not present initially (maybe to reconfirm them asking again to the user)

gdaniel commented:
Hi @mgv99,

Could this be handled by ContextParameterValue#getValue? You can use an object there that encapsulate the NER result as well as metadatas. Maybe this is not convenient, but could you tell me more about it?

I don't remember very well the status of the metamodel project, but I guess the "generated" code is not synchronized with the metamodel anymore, right? This would explain why the PR doesn't contain an update of the metamodel files.

mgv99 commented:
The thing is that from a state body, we access a parameter value with context.getIntent().getValue() (which contains typically a string value). Our idea is to get another attribute from the parameter with context.getIntent().getInfo(), that returns a Map with other things.

We considered your option, but we think that is better to create a dedicated attribute so the bot developer can forget about it if he/she wants and retrieve only the string value (it is typically a string).

Also, semantically it makes more sense to create a new attribute Info since what we want to store in it is not a part of the parameter value

gdaniel commented:
@jcabot right, that's what I did indeed. I think at some point I realised that we didn't need EMF at all, so I didn't bother with the EMF generation mechanism. I got confused because @mgv99 contributions look like EMF generated code, if you look at the [StateImpl](https://github.com/xatkit-bot-platform/xatkit-metamodels/blob/master/execution/src/main/java-gen/com/xatkit/execution/impl/StateImpl.java) class I didn't go this route and just added some attributes to the class and methods to access them instead of going through EMF's notification mechanism.

I'd suggest to do something like this: add a info map as an attribute to the `ContextParameterValue`, and a method to access it. Anyways, regardless of the solution we choose, you probably don't want to copy the `@Generated` tags in the documentation if the code is not generated.

@mgv99 I get your point, and it makes sense, thanks. Did you consider to name the attribute `metadata` or something similar? I guess its not a big deal but it seems more accurate to me.

gdaniel commented:
Yes, the main reason you don't have to mimick EMF code is that the model is purely a runtime artifact. You never store it as an XMI file, so you don't really care if some of the attributes aren't handled the EMF way.

mgv99 committed:
remove EMF code

The deleted code is supposed to be automatically generated by the xatkit metamodel. We decided to not mimic its code generation and just add the necessary attributes and methods. Added @NotGenerated tags

Changed the info data type from Object to Map<String, Object>

mgv99 commented:
Let me know if there is something missing/wrong in the last commit, thanks!
